core/src/services/mongodb/backend.rs

```diff
@@ -0,0 +1,315 @@
+    pub fn value_field(&mut self, value_field: &str) -> &mut Self {
+        if !value_field.is_empty() {
+            self.value_field = Some(value_field.to_string());
+        }
+        self
+    }
+}
+
+impl Builder for MongoDbBuilder {
+    const SCHEME: Scheme = Scheme::Mongodb;
+
+    type Accessor = MongoDbBackend;
+
+    fn from_map(map: std::collections::HashMap<String, String>) -> Self {
+        let mut builder = Self::default();
+
+        map.get("connection_string")
+            .map(|v| builder.connection_string(v));
+        map.get("database").map(|v| builder.database(v));
+        map.get("collection").map(|v| builder.collection(v));
+        map.get("root").map(|v| builder.root(v));
+        map.get("key_field").map(|v| builder.key_field(v));
+        map.get("value_field").map(|v| builder.value_field(v));
+
+        builder
+    }
+
+    fn build(&mut self) -> Result<Self::Accessor> {
+        let conn = match &self.connection_string.clone() {
+            Some(v) => v.clone(),
+            None => {
+                return Err(
+                    Error::new(ErrorKind::InvalidInput, "connection_string is required")
+                        .with_context("service", Scheme::Mongodb),
+                )
+            }
+        };
+        let database = match &self.database.clone() {
+            Some(v) => v.clone(),
+            None => {
+                return Err(Error::new(ErrorKind::InvalidInput, "database is required")
+                    .with_context("service", Scheme::Mongodb))
+            }
+        };
+        let collection = match &self.collection.clone() {
+            Some(v) => v.clone(),
+            None => {
+                return Err(
+                    Error::new(ErrorKind::InvalidInput, "collection is required")
+                        .with_context("service", Scheme::Mongodb),
+                )
+            }
+        };
+        let key_field = match &self.key_field.clone() {
+            Some(v) => v.clone(),
+            None => "key".to_string(),
+        };
+        let value_field = match &self.value_field.clone() {
+            Some(v) => v.clone(),
+            None => "value".to_string(),
+        };
+
+        Ok(MongoDbBackend::new(Adapter {
+            connection_string: conn,
+            client: OnceCell::new(),
+            options: OnceCell::new(),
+            database,
+            database_instance: OnceCell::new(),
+            collection,
+            collection_instance: OnceCell::new(),
+            key_field,
+            value_field,
+        }))
+    }
+}
+
+pub type MongoDbBackend = kv::Backend<Adapter>;
+
+#[derive(Clone)]
+pub struct Adapter {
+    connection_string: String,
+    client: OnceCell<Arc<mongodb::Client>>,
+    options: OnceCell<Arc<ClientOptions>>,
+    database: String,
+    database_instance: OnceCell<Arc<mongodb::Database>>,
+    collection: String,
+    collection_instance: OnceCell<Arc<mongodb::Collection<Document>>>,
```

Comment:
How about merging `client`, `options`, `database_instance` and `collection_instance` into one? `Collection` will store the client and database internally, so we don't need to store them manually.

And we don't need to make it inside `Arc` since `Collection` itself implements `Clone`

Reply:
I'm not sure about arc

In my thought, Arc and Clone are two different ways to solve different issue. Arc for thread safe reference and Clone for different instance in thread? Clone better than Arc?

Reply:
> I'm not sure about arc
> 
> In my thought, Arc and Clone are two different ways to solve different issue. Arc for thread safe reference and Clone for different instance in thread? Clone better than Arc?

Your understanding is correct. But `Collection` impls `Send + Sync + Clone`, so we don't need `Arc` to make it safe between threads. And internally, `Collection` contains an `Arc` for it's client/database references.